prompting/validators/reward/blacklist.py

```diff
@@ -16,35 +16,203 @@
+
+        Returns:
+            dict: Dictionary of n-gram tuples and their significance scores
+        """
+
+        significance_scores = {}
+        for ngram, count in self.counter.items():
+
+            # calculate significance score for ngram
+            significance_scores[ngram] = self.A ** (len(ngram) - 1) * (count - 1)
+
+        self._last_update = self._running_size
+
+        return significance_scores
+
+    def get_significance(self) -> dict:
+        """Get significance scores, only recalculating if the counter has been updated.
+
+        Returns:
+            dict: Dictionary of n-gram tuples and their significance scores
+        """
+
+        if self._last_update != self._running_size:
+            self.significance_scores = self.calculate_significance()
+
+        return self.significance_scores
+
+    def most_common(self, n:int = 10) -> dict:
+        """Get most common n-grams in queue
+
+        Args:
+            n (int): Number of most common n-grams to return. Defaults to 10.
+
+        Returns:
+            dict: Sorted dictionary of n-gram tuples and their counts
+        """
+        return self.counter.most_common(n)
+
+    def most_significant(self, n:int = 10, force_update:bool = True) -> dict:
+        """Get most significant n-grams in queue based on significance scores
+
+        Args:
+            n (int, optional): Number of most significant n-grams to return. Defaults to 10.
+            force_update (bool, optional): Force recalculate the significance scores. Defaults to True.
+
+        Returns:
+            dict: Sorted dictionary of n-gram tuples and their significance scores
+        """
+
+
+        scores = self.get_significance() if force_update else self.significance_scores
+
+        return sorted(scores.items(), key=lambda x: x[1], reverse=True)[:n]
 
     def reward(self, prompt: str, completion: str, name: str) -> float:
-        if completion in blacklist:
-            return 0.0
+        """Reward function for blacklist reward model. Returns 1 if completion contains an n-gram with significance above the boundary, 0 otherwise.
 
-        if completion == prompt:
-            return 0.0
+        Args:
+            prompt (str): Prompt text
+            completion (str): Completion text
+            name (str): Name of the validation step
+
+        Returns:
+            float: Reward value {0,1}
+        """
 
-        if completion in self.question_blacklist or completion in self.answer_blacklist:
+        if completion in prompt:
             return 0.0
 
-        return 1
+        # Extract n-grams from completion
+        ngrams = self.extract_ngrams(completion.lower())
+        # Get significance scores
+        scores = self.get_significance()
+
+        # Check if any n-grams have significance above the boundary
+        for ngram in ngrams:
+            if scores.get(ngram) > self.boundary:
```

Comment:
It would be of interest to include a `self.blacklist` attribute which contains all the ngrams with `significance>self.boundary`. It should not be a very large object (10s-100s items). This is something we could log to wandb regularly to introspect the blacklist.